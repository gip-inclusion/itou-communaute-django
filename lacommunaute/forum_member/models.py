from django.db import models
from machina.apps.forum_member.abstract_models import AbstractForumProfile

from lacommunaute.forum_member.enums import ActiveSearch, Regions
from lacommunaute.forum_member.shortcuts import get_forum_member_display_name


class ForumProfile(AbstractForumProfile):
    linkedin = models.URLField(blank=True, null=True, verbose_name="lien vers votre profil LinkedIn")
    cv = models.FileField(upload_to="cv/", blank=True, null=True, verbose_name="votre CV")
    search = models.CharField(
        max_length=20,
        blank=True,
        null=True,
        choices=ActiveSearch.choices,
        default=ActiveSearch.NO,
        verbose_name="recherche active",
    )
    region = models.CharField(
        max_length=100, blank=True, null=True, choices=Regions.choices, default=Regions.RXX, verbose_name="région"
    )
    internship_duration = models.IntegerField(default=0, verbose_name="durée du stage (en mois)")
    updated_at = models.DateTimeField(auto_now=True)

    def __str__(self):
        return get_forum_member_display_name(self.user)
