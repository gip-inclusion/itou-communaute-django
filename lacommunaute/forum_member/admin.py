from django.contrib import admin
from django.db.models import Q
from django.utils.html import format_html
from machina.apps.forum_member.admin import ForumProfileAdmin
from machina.core.db.models import get_model


ForumProfile = get_model("forum_member", "ForumProfile")


class HasCVFilter(admin.SimpleListFilter):
    title = "A un CV"
    parameter_name = "has_cv"

    def lookups(self, request, model_admin):
        return (("yes", "Oui"), ("no", "Non"))

    def queryset(self, request, queryset):
        conditions = Q(cv__isnull=True) | Q(cv="")
        value = self.value()
        if value == "yes":
            return queryset.exclude(conditions)
        if value == "no":
            return queryset.filter(conditions)
        return queryset


class HasLinkedInFilter(admin.SimpleListFilter):
    title = "A un lien LinkedIn"
    parameter_name = "has_linkedin"

    def lookups(self, request, model_admin):
        return (("yes", "Oui"), ("no", "Non"))

    def queryset(self, request, queryset):
        value = self.value()
        conditions = Q(linkedin__isnull=True) | Q(linkedin="")
        if value == "yes":
            return queryset.exclude(conditions)
        if value == "no":
            return queryset.filter(conditions)
        return queryset


class CommuForumProfileAdmin(ForumProfileAdmin):
    list_display = ("id", "user", "search", "region", "cv_display", "linkedin_display")
    list_filter = (
        "search",
        HasCVFilter,
        HasLinkedInFilter,
        "region",
    )

    @admin.display(description="CV")
    def cv_display(self, instance):
        if instance.cv:
            return format_html(
                "<a href='{}' target='_blank'>Télécharger le CV</a>",
                instance.cv.url,
            )

    @admin.display(description="LinkedIn")
    def linkedin_display(self, instance):
        if instance.linkedin:
            return format_html('<a href="{}">Lien du profil LK</a>', instance.linkedin)


admin.site.unregister(ForumProfile)
admin.site.register(ForumProfile, CommuForumProfileAdmin)
