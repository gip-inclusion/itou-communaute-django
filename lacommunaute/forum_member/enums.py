from django.db import models


class ActiveSearch(models.TextChoices):
    NO = "NO", "Je ne suis pas en recherche"
    Internship = "INTERNSHIP", "Je suis en recherche active de stage de CIP"
    Apprenticeship = "APPRENTICESHIP", "Je suis en recherche active d'apprentissage de CIP"


class Regions(models.TextChoices):
    RXX = "XX", "---------"
    R00 = "00", "France entière"
    R84 = "84", "Auvergne-Rhône-Alpes"
    R27 = "27", "Bourgogne-Franche-Comté"
    R53 = "53", "Bretagne"
    R24 = "24", "Centre-Val de Loire"
    R94 = "94", "Corse"
    R44 = "44", "Grand Est"
    R01 = "01", "Guadeloupe"
    R03 = "03", "Guyane"
    R32 = "32", "Hauts-de-France"
    R11 = "11", "Île-de-France"
    R04 = "04", "La Réunion"
    R02 = "02", "Martinique"
    R06 = "06", "Mayotte"
    R28 = "28", "Normandie"
    R75 = "75", "Nouvelle-Aquitaine"
    R76 = "76", "Occitanie"
    R52 = "52", "Pays de la Loire"
    R93 = "93", "Provence-Alpes-Côte d’Azur"
