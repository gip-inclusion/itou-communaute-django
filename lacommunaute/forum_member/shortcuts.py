"""
    Forum member shortcuts
    ======================

    This module defines shortcut functions allowing to easily get forum member information.

"""


def get_forum_member_display_name(user):
    lastname = ""
    firstname = ""

    if user.last_name:
        lastname = f"{user.last_name.capitalize()[0]}."

    if user.first_name:
        firstname = user.first_name.capitalize()

    return " ".join([firstname, lastname]).strip()
