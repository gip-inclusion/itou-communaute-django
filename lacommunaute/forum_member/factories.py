import factory

from lacommunaute.forum_member.models import ForumProfile
from lacommunaute.users.factories import UserFactory


class ForumProfileFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = ForumProfile

    user = factory.SubFactory(UserFactory)
    signature = factory.Faker("sentence", nb_words=40)
