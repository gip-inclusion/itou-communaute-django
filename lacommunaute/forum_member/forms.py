from django import forms
from machina.apps.forum_member.forms import ForumProfileForm as BaseForumProfileForm

from lacommunaute.forum_member.enums import ActiveSearch, Regions
from lacommunaute.forum_member.models import ForumProfile


class ForumProfileForm(BaseForumProfileForm):
    cv = forms.FileField(label="Curriculum Vitae", required=False)
    linkedin = forms.URLField(label="Lien vers votre profil LinkedIn", required=False)
    search = forms.ChoiceField(label="Je suis en recherche active", choices=ActiveSearch.choices, required=False)
    region = forms.ChoiceField(label="Région", choices=Regions.choices, required=False)
    internship_duration = forms.IntegerField(label="Durée du stage (en mois)", required=False)

    class Meta:
        model = ForumProfile
        fields = ["avatar", "signature", "linkedin", "cv", "search", "region", "internship_duration"]
