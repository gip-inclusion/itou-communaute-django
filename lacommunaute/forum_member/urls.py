from django.urls import path

from lacommunaute.forum_member.views import (
    ForumProfileDetailView,
    ForumProfileUpdateView,
    JoinForumFormView,
    JoinForumLandingView,
    ModeratorProfileListView,
)


app_name = "members"

urlpatterns = [
    path("profile/edit/", ForumProfileUpdateView.as_view(), name="profile_update"),
    path("profile/<str:username>/", ForumProfileDetailView.as_view(), name="profile"),
    path("forum/<str:slug>-<int:pk>/", ModeratorProfileListView.as_view(), name="forum_profiles"),
    path("join-forum-landing/<uuid:token>/", JoinForumLandingView.as_view(), name="join_forum_landing"),
    path("join-forum/<uuid:token>/", JoinForumFormView.as_view(), name="join_forum_form"),
]
