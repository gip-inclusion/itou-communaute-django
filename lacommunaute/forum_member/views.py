import logging

from django.contrib.auth.mixins import LoginRequiredMixin
from django.shortcuts import get_object_or_404
from django.urls import reverse
from django.utils.http import urlencode
from django.views.generic import FormView, ListView, TemplateView
from machina.apps.forum_member.views import (
    ForumProfileDetailView as BaseForumProfileDetailView,
    ForumProfileUpdateView as BaseForumProfileUpdateView,
)
from machina.core.loading import get_class

from lacommunaute.forum.models import Forum
from lacommunaute.forum_member.forms import ForumProfileForm, JoinForumForm
from lacommunaute.forum_member.models import ForumProfile
from lacommunaute.utils.urls import get_safe_url


logger = logging.getLogger(__name__)


PermissionRequiredMixin = get_class("forum_permission.viewmixins", "PermissionRequiredMixin")


class ForumProfileDetailView(BaseForumProfileDetailView):
    slug_field = "username"
    slug_url_kwarg = "username"


class ForumProfileUpdateView(BaseForumProfileUpdateView):
    form_class = ForumProfileForm

    def get_success_url(self):
        return reverse("members:profile", kwargs={"username": self.request.user.username})


class ModeratorProfileListView(PermissionRequiredMixin, ListView):
    model = ForumProfile
    template_name = "forum_member/moderator_profiles.html"
    context_object_name = "forum_profiles"
    paginate_by = 60
    permission_required = ["can_approve_posts"]

    def get_forum(self):
        """Returns the forum to consider."""
        if not hasattr(self, "forum"):
            self.forum = get_object_or_404(Forum, pk=self.kwargs["pk"])
        return self.forum

    def get_queryset(self):
        """Returns the list of items for this view."""
        self.forum = self.get_forum()
        users = self.forum.members_group.user_set.all()
        return ForumProfile.objects.filter(user__in=users).select_related("user").order_by("user__first_name")

    def get_controlled_object(self):
        """Returns the controlled object."""
        return self.get_forum()

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        forum = self.get_forum()
        context["forum"] = forum
        return context


class JoinForumLandingView(TemplateView):
    template_name = "forum_member/join_forum_landing.html"

    def get_forum(self):
        if not hasattr(self, "forum"):
            self.forum = get_object_or_404(
                Forum,
                invitation_token=self.kwargs["token"],
            )
        return self.forum

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        forum = self.get_forum()
        context["forum"] = forum
        params = {
            "previous_url": reverse(
                "members:join_forum_landing",
                kwargs={
                    "token": forum.invitation_token,
                },
            ),
            "next_url": get_safe_url(self.request, "next"),
        }
        context["inclusion_connect_url"] = f"{reverse('inclusion_connect:authorize')}?{urlencode(params)}"
        return context


class JoinForumFormView(LoginRequiredMixin, FormView):
    template_name = "forum_member/join_forum_form.html"
    form_class = JoinForumForm

    def get_forum(self):
        if not hasattr(self, "forum"):
            self.forum = get_object_or_404(
                Forum,
                invitation_token=self.kwargs["token"],
            )
        return self.forum

    def form_valid(self, form):
        form.forum = self.get_forum()
        form.user = self.request.user
        form.join_forum()
        return super().form_valid(form)

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["forum"] = self.get_forum()
        return context

    def get_login_url(self):
        return reverse(
            "members:join_forum_landing",
            kwargs={
                "token": self.kwargs["token"],
            },
        )

    def get_success_url(self):
        return reverse(
            "forum_extension:forum",
            kwargs={
                "slug": self.forum.slug,
                "pk": self.forum.pk,
            },
        )


class LeaderBoardListView(ListView):
    model = ForumProfile
    template_name = "forum_member/profiles.html"
    context_object_name = "forum_profiles"
    paginate_by = 78

    def get_queryset(self):
        return ForumProfile.objects.power_users()

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["subtitle"] = "Contributeurs authentifi√©s les plus actifs sur les 30 derniers jours"
        return context


class SeekersListView(ListView):
    model = ForumProfile
    template_name = "forum_member/seekers_profiles.html"
    context_object_name = "forum_profiles"
    paginate_by = 78

    def get_queryset(self):
        return ForumProfile.objects.exclude(search="NO")

    def get_context_data(self, **kwargs):
        context = super().get_context_data(**kwargs)
        context["subtitle"] = "CIP en recherche active de stage ou d'alternance"
        return context
