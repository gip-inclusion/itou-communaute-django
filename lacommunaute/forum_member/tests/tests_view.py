import pytest
from django.test import RequestFactory, TestCase
from django.urls import reverse
from machina.core.loading import get_class
from pytest_django.asserts import assertContains, assertNotContains

from lacommunaute.forum_member.factories import ForumProfileFactory
from lacommunaute.forum_member.models import ForumProfile
from lacommunaute.forum_member.shortcuts import get_forum_member_display_name
from lacommunaute.forum_member.views import ForumProfileUpdateView
from lacommunaute.users.factories import UserFactory
from lacommunaute.utils.testing import parse_response_to_soup


PermissionHandler = get_class("forum_permission.handler", "PermissionHandler")
assign_perm = get_class("forum_permission.shortcuts", "assign_perm")


class ForumProfileUpdateViewTest(TestCase):
    def test_success_url(self):
        forum_profiles = ForumProfileFactory()
        view = ForumProfileUpdateView()
        view.request = RequestFactory().get("/")
        view.request.user = forum_profiles.user
        self.assertEqual(
            view.get_success_url(), reverse("members:profile", kwargs={"username": forum_profiles.user.username})
        )

    def test_form_fields(self):
        forum_profiles = ForumProfileFactory()
        self.client.force_login(forum_profiles.user)
        response = self.client.get(reverse("members:profile_update"))

        self.assertContains(response, "linkedin", status_code=200)
        self.assertContains(response, "cv")
        self.assertContains(response, "search")
        self.assertContains(response, "region")
        self.assertContains(response, "internship_duration")


class TestSeekersListView:
    def test_content(self, client, db):
        undesired_forum_profile = ForumProfileFactory()
        internship_forum_profile = ForumProfileFactory(search="INTERNSHIP")
        apprentice_forum_profile = ForumProfileFactory(search="APPRENTICESHIP")

        response = client.get(reverse("members:seekers"))
        assertContains(response, get_forum_member_display_name(internship_forum_profile.user))
        assertContains(response, get_forum_member_display_name(apprentice_forum_profile.user))
        assertNotContains(response, get_forum_member_display_name(undesired_forum_profile.user))
        assert response.context_data["subtitle"] == "CIP en recherche active de stage ou d'alternance"

    def test_sorted_profils(self, client, db):
        ForumProfileFactory(search="APPRENTICESHIP")
        ForumProfileFactory(search="INTERNSHIP")
        response = client.get(reverse("members:seekers"))
        # test queryset is ordered by updated_at
        assert response.context_data["forum_profiles"][0] == ForumProfile.objects.last()
        assert response.context_data["forum_profiles"][1] == ForumProfile.objects.first()

    @pytest.mark.parametrize("authenticated,snapshot_name", [(True, "authenticated_user"), (False, "anonymous_user")])
    def test_update_profil_link(self, client, db, authenticated, snapshot_name, snapshot):
        if authenticated:
            client.force_login(UserFactory())
        response = client.get(reverse("members:seekers"))
        content = parse_response_to_soup(response, selector="#action-box")
        assert str(content) == snapshot(name=snapshot_name)


class TestForumProfileDetailView:
    def test_show_linkedin_link(self, client, db):
        forum_profile = ForumProfileFactory(linkedin="https://www.linkedin.com/in/johndoe/")
        response = client.get(reverse("members:profile", kwargs={"username": forum_profile.user.username}))
        assertContains(response, forum_profile.linkedin)
