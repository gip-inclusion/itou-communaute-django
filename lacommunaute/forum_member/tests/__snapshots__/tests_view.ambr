# serializer version: 1
# name: TestSeekersListView.test_update_profil_link[False-anonymous_user][anonymous_user]
  '''
  <section class="s-section" id="action-box">
          <div class="s-section__container container">
              
                  <div class="s-section__row row">
                      <div class="s-section__col col-12">
                          <div class="c-box mb-5">
                              <div class="row">
                                  <div class="col-lg col-12 mb-3">Je me connecte pour mettre à jour mon profil</div>
                                  <div class="col-lg-auto col-12">
  
  
      <a class="btn btn-outline-primary btn-ico btn-block" href="/users/login/?next=%2Fmembers%2Fprofile%2Fedit%2F">Se connecter | S'inscrire</a>
  
  </div>
                              </div>
                          </div>
                      </div>
                  </div>
              
              <div class="s-section__row row">
                  <div class="s-section__col col-12">
                      <div class="row row-cols-1 row-cols-md-2 g-4">
                          
                              <div class="col">
                                  <p>No body yet…</p>
                              </div>
                          
                      </div>
                      
                  </div>
              </div>
          </div>
      </section>
  '''
# ---
# name: TestSeekersListView.test_update_profil_link[True-authenticated_user][authenticated_user]
  '''
  <section class="s-section" id="action-box">
          <div class="s-section__container container">
              
                  <div class="s-section__row row">
                      <div class="s-section__col col-12">
                          <div class="c-box mb-5">
                              <div class="row">
                                  <div class="col-lg col-12 mb-3">
                                      <p>Je suis CIP en recherche d'un stage ou d'une alternance ? Je complète mon profil pour me faire connaitre.</p>
                                  </div>
                                  <div class="col-lg-auto col-12">
                                      <a class="btn btn-primary" href="/members/profile/edit/">Mettre à jour mon profil</a>
                                  </div>
                              </div>
                          </div>
                      </div>
                  </div>
              
              <div class="s-section__row row">
                  <div class="s-section__col col-12">
                      <div class="row row-cols-1 row-cols-md-2 g-4">
                          
                              <div class="col">
                                  <p>No body yet…</p>
                              </div>
                          
                      </div>
                      
                  </div>
              </div>
          </div>
      </section>
  '''
# ---
